File: plugin/src/net/sf/orcc/network/parser/ExprParser.java
/*
 * Copyright (c) 2009, IETR/INSA of Rennes
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * 
 *   * Redistributions of source code must retain the above copyright notice,
 *     this list of conditions and the following disclaimer.
 *   * Redistributions in binary form must reproduce the above copyright notice,
 *     this list of conditions and the following disclaimer in the documentation
 *     and/or other materials provided with the distribution.
 *   * Neither the name of the IETR/INSA of Rennes nor the names of its
 *     contributors may be used to endorse or promote products derived from this
 *     software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
 * WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */
package net.sf.orcc.network.parser;

import java.util.ArrayList;
import java.util.List;

import net.sf.orcc.OrccException;
import net.sf.orcc.common.LocalVariable;
import net.sf.orcc.common.Location;
import net.sf.orcc.ir.actor.VarUse;
import net.sf.orcc.ir.expr.BinaryOp;
import net.sf.orcc.ir.expr.BooleanExpr;
import net.sf.orcc.ir.expr.IExpr;
import net.sf.orcc.ir.expr.IntExpr;
import net.sf.orcc.ir.expr.ListExpr;
import net.sf.orcc.ir.expr.StringExpr;
import net.sf.orcc.ir.expr.UnaryExpr;
import net.sf.orcc.ir.expr.UnaryOp;
import net.sf.orcc.ir.expr.VarExpr;

import org.w3c.dom.Element;
import org.w3c.dom.Node;

/**
 * This class defines a parser of XDF expressions.
 * 
 * @author Matthieu Wipliez
 * 
 */
public class ExprParser {

	/**
	 * Parses the given node as an expression and returns the matching IExpr
	 * expression.
	 * 
	 * @param node
	 *            a node whose expected to be, or whose sibling is expected to
	 *            be, a DOM element named "Expr".
	 * @return an expression
	 * @throws OrccException
	 *             if the given node or its siblings could not be parsed as an
	 *             expression
	 */
	public IExpr parseExpr(Node node) throws OrccException {
		ParseContinuation<IExpr> cont = parseExprCont(node);
		IExpr expr = cont.getResult();
		if (expr == null) {
			throw new OrccException("Expected an Expr element");
		} else {
			return expr;
		}
	}

	/**
	 * Parses the given node as a binary operator and returns a parse
	 * continuation with the operator parsed.
	 * 
	 * @param node
	 *            a node that is expected, or whose sibling is expected, to be a
	 *            DOM element named "Op".
	 * @return a parse continuation with the operator parsed
	 * @throws OrccException
	 *             if the binary operator could not be parsed
	 */
	private ParseContinuation<BinaryOp> parseExprBinaryOp(Node node)
			throws OrccException {
		while (node != null) {
			if (node.getNodeName().equals("Op")) {
				Element op = (Element) node;
				String name = op.getAttribute("name");
				if (name.equals("and")) {
					return new ParseContinuation<BinaryOp>(node, BinaryOp.LAND);
				} else if (name.equals("/")) {
					return new ParseContinuation<BinaryOp>(node, BinaryOp.DIV);
				} else if (name.equals("div")) {
					return new ParseContinuation<BinaryOp>(node,
							BinaryOp.DIV_INT);
				} else if (name.equals("=")) {
					return new ParseContinuation<BinaryOp>(node, BinaryOp.EQ);
				} else if (name.equals("^")) {
					return new ParseContinuation<BinaryOp>(node, BinaryOp.EXP);
				} else if (name.equals(">=")) {
					return new ParseContinuation<BinaryOp>(node, BinaryOp.GE);
				} else if (name.equals(">")) {
					return new ParseContinuation<BinaryOp>(node, BinaryOp.GT);
				} else if (name.equals("<=")) {
					return new ParseContinuation<BinaryOp>(node, BinaryOp.LE);
				} else if (name.equals("<")) {
					return new ParseContinuation<BinaryOp>(node, BinaryOp.LT);
				} else if (name.equals("-")) {
					return new ParseContinuation<BinaryOp>(node, BinaryOp.MINUS);
				} else if (name.equals("mod")) {
					return new ParseContinuation<BinaryOp>(node, BinaryOp.MOD);
				} else if (name.equals("!=")) {
					return new ParseContinuation<BinaryOp>(node, BinaryOp.NE);
				} else if (name.equals("or")) {
					return new ParseContinuation<BinaryOp>(node, BinaryOp.LOR);
				} else if (name.equals("+")) {
					return new ParseContinuation<BinaryOp>(node, BinaryOp.PLUS);
				} else if (name.equals("*")) {
					return new ParseContinuation<BinaryOp>(node, BinaryOp.TIMES);
				} else {
					throw new OrccException("Unknown binary operator \"" + name
							+ "\"");
				}
			}

			node = node.getNextSibling();
		}

		return new ParseContinuation<BinaryOp>(node, null);
	}

	/**
	 * Parses the given node and its siblings as a sequence of binary
	 * operations, aka "BinOpSeq". A BinOpSeq is a sequence of expr, op, expr,
	 * op, expr...
	 * 
	 * @param node
	 *            the first child node of a Expr kind="BinOpSeq" element
	 * @return a parse continuation with a BinaryExpr
	 * @throws OrccException
	 *             if something goes wrong
	 */
	private ParseContinuation<IExpr> parseExprBinOpSeq(Node node)
			throws OrccException {
		List<Object> args = new ArrayList<Object>();

		ParseContinuation<? extends Object> cont = parseExprCont(node);
		args.add(cont.getResult());
		node = cont.getNode();
		while (node != null) {
			cont = parseExprBinaryOp(node);
			BinaryOp op = (BinaryOp) cont.getResult();
			node = cont.getNode();
			if (op != null) {
				args.add(op);

				cont = parseExprCont(node);
				IExpr expr = (IExpr) cont.getResult();
				if (expr == null) {
					throw new OrccException("Expected an Expr element");
				}

				args.add(expr);
				node = cont.getNode();
			}
		}

		// TODO one more time, apply operator priority to stupid binary
		// operation sequence...
		return new ParseContinuation<IExpr>(node, new IntExpr(42));
	}

	/**
	 * Parses the given node as an expression and returns the matching IExpr
	 * expression.
	 * 
	 * @param node
	 *            a node whose sibling is expected to be a DOM element named
	 *            "Expr".
	 * @return an expression
	 * @throws OrccException
	 *             if the given node or its siblings could not be parsed as an
	 *             expression
	 */
	private ParseContinuation<IExpr> parseExprCont(Node node)
			throws OrccException {
		IExpr expr = null;
		while (node != null) {
			if (node.getNodeName().equals("Expr")) {
				Element elt = (Element) node;
				String kind = elt.getAttribute("kind");
				if (kind.equals("BinOpSeq")) {
					return parseExprBinOpSeq(elt.getFirstChild());
				} else if (kind.equals("Literal")) {
					expr = parseExprLiteral(elt);
					break;
				} else if (kind.equals("List")) {
					List<IExpr> exprs = parseExprs(node.getFirstChild());
					expr = new ListExpr(new Location(), exprs);
					break;
				} else if (kind.equals("UnaryOp")) {
					ParseContinuation<UnaryOp> cont = parseExprUnaryOp(node
							.getFirstChild());
					UnaryOp op = cont.getResult();
					IExpr unaryExpr = parseExpr(cont.getNode());
					expr = new UnaryExpr(new Location(), op, unaryExpr, null);
					break;
				} else if (kind.equals("Var")) {
					String name = elt.getAttribute("name");
					LocalVariable varDef = new LocalVariable(false, false, 0,
							null, name, null, null, null, null);
					VarUse varUse = new VarUse(varDef, null);
					expr = new VarExpr(new Location(), varUse);
					break;
				} else {
					throw new OrccException("Unsupported Expr kind: \"" + kind
							+ "\"");
				}
			}

			node = node.getNextSibling();
		}

		return new ParseContinuation<IExpr>(node, expr);
	}

	/**
	 * Parses the given "Expr" element as a literal and returns the matching
	 * IExpr expression.
	 * 
	 * @param elt
	 *            a DOM element named "Expr"
	 * @return an expression
	 * @throws OrccException
	 *             if the literal could not be parsed
	 */
	private IExpr parseExprLiteral(Element elt) throws OrccException {
		String kind = elt.getAttribute("literal-kind");
		String value = elt.getAttribute("value");
		if (kind.equals("Boolean")) {
			return new BooleanExpr(new Location(), Boolean.parseBoolean(value));
		} else if (kind.equals("Character")) {
			throw new OrccException("Characters not supported yet");
		} else if (kind.equals("Integer")) {
			return new IntExpr(new Location(), Integer.parseInt(value));
		} else if (kind.equals("Real")) {
			throw new OrccException("Reals not supported yet");
		} else if (kind.equals("String")) {
			return new StringExpr(new Location(), value);
		} else {
			throw new OrccException("Unsupported Expr " + "literal kind: \""
					+ kind + "\"");
		}
	}

	private List<IExpr> parseExprs(Node node) throws OrccException {
		List<IExpr> exprs = new ArrayList<IExpr>();
		while (node != null) {
			if (node.getNodeName().equals("Expr")) {
				exprs.add(parseExpr(node));
			}

			node = node.getNextSibling();
		}

		return exprs;
	}

	/**
	 * Parses the given node as a unary operator and returns a parse
	 * continuation with the operator parsed.
	 * 
	 * @param node
	 *            a node that is expected, or whose sibling is expected, to be a
	 *            DOM element named "Op".
	 * @return a parse continuation with the operator parsed
	 * @throws OrccException
	 *             if the unary operator could not be parsed
	 */
	private ParseContinuation<UnaryOp> parseExprUnaryOp(Node node)
			throws OrccException {
		while (node != null) {
			if (node.getNodeName().equals("Op")) {
				Element op = (Element) node;
				String name = op.getAttribute("name");
				if (name.equals("#")) {
					return new ParseContinuation<UnaryOp>(node,
							UnaryOp.NUM_ELTS);
				} else if (name.equals("not")) {
					return new ParseContinuation<UnaryOp>(node, UnaryOp.LNOT);
				} else if (name.equals("-")) {
					return new ParseContinuation<UnaryOp>(node, UnaryOp.MINUS);
				} else {
					throw new OrccException("Unknown unary operator \"" + name
							+ "\"");
				}
			}

			node = node.getNextSibling();
		}

		throw new OrccException("Expected an Op element");
	}

}

File: projects/Jade/lib/XDFParser/ExprParser.cpp
/*
 * Copyright (c) 2009, IETR/INSA of Rennes
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * 
 *   * Redistributions of source code must retain the above copyright notice,
 *     this list of conditions and the following disclaimer.
 *   * Redistributions in binary form must reproduce the above copyright notice,
 *     this list of conditions and the following disclaimer in the documentation
 *     and/or other materials provided with the distribution.
 *   * Neither the name of the IETR/INSA of Rennes nor the names of its
 *     contributors may be used to endorse or promote products derived from this
 *     software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
 * WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/**
@brief Implementation of ExprParser
@author Jerome Gorin
@file ExprParser.cpp
@version 0.1
@date 2010/04/12
*/

//------------------------------
#include <iostream>
#include <string>
#include <list>

#include "llvm/Constants.h"
#include "llvm/DerivedTypes.h"
#include "llvm/LLVMContext.h"

#include "BinOpSeqParser.h"
#include "ExprParser.h"
#include "Jade/Core/Expr/BoolExpr.h"
#include "Jade/Core/Expr/IntExpr.h"
#include "Jade/Core/Expr/BinaryExpr.h"
#include "Jade/Core/Expr/StringExpr.h"
//------------------------------

using namespace std;
using namespace llvm;

ExprParser::ExprParser (llvm::LLVMContext& C): Context(C){
}

ExprParser::~ExprParser (){

}

Expr* ExprParser::parseExpr(TiXmlNode* node){
	ParseContinuation<Expr*> cont = parseExprCont(node);

	Expr* expr = cont.getResult();

	if (expr == NULL) {
		cerr << "Expected an expression element";
		exit(0);
	} 

	return expr;	
}

ParseContinuation<Expr*> ExprParser::parseExprCont(TiXmlNode* node){
	Expr* expression = NULL;

	while(node != NULL){
		if (TiXmlString(node->Value()) == XDFNetwork::EXPR) {
			TiXmlElement* elt = (TiXmlElement*)node;
			TiXmlString kind(elt->Attribute(XDFNetwork::KIND));

			if (kind == XDFNetwork::KIND_BINOPSEQ) {
				return parseExprBinOpSeq(node->FirstChild());
			} else if (kind == XDFNetwork::KIND_LITERAL) {
				expression = parseExprLiteral(elt);
				break;
			} else if (kind == XDFNetwork::KIND_LIST) {
				cerr << "List not supported yet";
				exit(0);
			} else if (kind == XDFNetwork::KIND_UNARYOP) {
				cerr << "UnaryOp not supported yet";
				exit(0);
			} else if (kind == XDFNetwork::KIND_VAR) {
				cerr << "Var not supported yet";
				exit(0);
			} else {
				cerr << "Unsupported expression kind: "<< kind.c_str();
				exit(0);
			}
		}
		
		node = node->NextSibling();	
	}

	return ParseContinuation<Expr*>(node, expression);
}

Expr* ExprParser::parseExprLiteral(TiXmlElement* elt){
	TiXmlString kind(elt->Attribute(XDFNetwork::LITERAL_KIND));
	TiXmlString value(elt->Attribute(XDFNetwork::LITERAL_VALUE));
	
	if (kind == XDFNetwork::LITERAL_BOOL) {
		return parseBoolean(value);	
	} else if (kind == XDFNetwork::LITERAL_CHAR) {
		fprintf(stderr,"Characters not supported yet");
		exit(0);
	} else if (kind == XDFNetwork::LITERAL_INT) {
		int integer = atoi (value.c_str());
		if (integer == 0 && value != "0"){
			fprintf(stderr,"Expression is not an integer");
			exit(0);
		}
		return new IntExpr(Context, integer);
	} else if (kind == XDFNetwork::LITERAL_REAL) {
		fprintf(stderr,"Reals not supported yet");
		exit(0);
	} else if (kind == XDFNetwork::LITERAL_STRING) {
		fprintf(stderr,"String not supported yet");
		exit(0);
	} else {
		fprintf(stderr,"Unsupported expression literal kind: \"%s\"", kind);
		exit(0);
	}
}


ParseContinuation<Expr*> ExprParser::parseExprBinOpSeq(TiXmlNode* node){
	list<Expr*> exprs;
	list<BinaryOp*> ops;

	ParseContinuation<Expr*> contE = parseExprCont(node);
	exprs.push_back(contE.getResult());
	node = contE.getNode();

	while (node != NULL){
		ParseContinuation<BinaryOp*> contO = parseExprBinaryOp(node);
		BinaryOp* op = contO.getResult();
		node = contO.getNode();

		if (op != NULL) {
			ops.push_back(op);

			contE = parseExprCont(node);
			Expr* expr = contE.getResult();
			if (expr == NULL) {
				cerr << "Missing an Expr element in Network.";
				exit(1);
			}

			exprs.push_back(expr);
			node = contE.getNode();
		}

	}

	Expr* expr = BinOpSeqParser().parse(Context, &exprs, &ops);
	return ParseContinuation<Expr*>(node, expr);
}

Expr* ExprParser::parseBoolean(TiXmlString value){
	if (value == "true") {
		return new BoolExpr(Context, true);
	}else if (value =="false") {
		return new BoolExpr(Context, false);
	}
	
	fprintf(stderr,"Expected a boolean value");
	exit(0);
}

ParseContinuation<BinaryOp*> ExprParser::parseExprBinaryOp(TiXmlNode* node){
	while (node != NULL){
		if (TiXmlString(node->Value()) == XDFNetwork::EXPR_OP) {
			TiXmlElement* op = (TiXmlElement*)node;
			string opName = op->Attribute(XDFNetwork::NAME);
			return ParseContinuation<BinaryOp*>(node, new BinaryOp(opName));
		}
		node = node->NextSibling();
	}
	return ParseContinuation<BinaryOp*>(node, NULL);
}

